redfish/collector.go

```diff
@@ -0,0 +1,241 @@
+package redfish
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/setup-hw/config"
+	"github.com/cybozu-go/setup-hw/gabs"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const namespace = "hw"
+
+// Collector implements prometheus.Collector interface.
+type Collector struct {
+	rules  []convertRule
+	client *client
+	cache  *cache
+}
+
+// CollectorConfig is a set of configurations for Collector.
+type CollectorConfig struct {
+	AddressConfig *config.AddressConfig
+	Port          string
+	UserConfig    *config.UserConfig
+	Rule          []byte
+}
+
+// NewCollector returns a new instance of Collector.
+func NewCollector(cc *CollectorConfig) (*Collector, error) {
```

Comment:
return `prometheus.Collector`.

Reply:
Returning `prometheus.Collector` causes an error in prom.go because we additionally implemented `Update()` for our collector.

```
collector, err := redfish.NewCollector(cc)
...
collector.Update(ctx, opts.redfishRoot)
```

=> `collector.Update undefined (type prometheus.Collector has no field or method Update)`

Reply:
I see.